src/sage/rings/localization.py

```diff
@@ -34,123 +34,123 @@
     [5 0 0]
     [0 3 0]
     [0 0 3]
-    sage: mFp1.base_ring()
+    sage: mFp1.base_ring()                                                              # optional - sage.modules sage.rings.finite_rings
     Finite Field of size 17
-    sage: mFp2 = matrix({k:f(v) for k, v in m2.dict().items()}); mFp2
+    sage: mFp2 = matrix({k: f(v) for k, v in m2.dict().items()}); mFp2                  # optional - sage.modules sage.rings.finite_rings
     [ 2  3  0]
     [ 9  8  0]
     [ 0  0 16]
-    sage: mFp3 = matrix({k:f(v) for k, v in m3.dict().items()}); mFp3
+    sage: mFp3 = matrix({k: f(v) for k, v in m3.dict().items()}); mFp3                  # optional - sage.modules sage.rings.finite_rings
     [16  0  0]
     [ 0  4  5]
     [ 0  7  6]
 
 Obtain specializations in characteristic 0::
 
-    sage: fQ = L.hom((3,5,7,11), codomain=QQ); fQ
+    sage: fQ = L.hom((3,5,7,11), codomain=QQ); fQ                                       # optional - sage.rings.finite_rings
     Ring morphism:
-      From: Multivariate Polynomial Ring in u0, u1, u2, q over Integer Ring localized at
-            (q, q + 1, u2, u1, u1 - u2, u0, u0 - u2, u0 - u1, u2*q - u1, u2*q - u0,
-            u1*q - u2, u1*q - u0, u0*q - u2, u0*q - u1)
+      From: Multivariate Polynomial Ring in u0, u1, u2, q over Integer Ring
+            localized at (q, q + 1, u2, u1, u1 - u2, u0, u0 - u2, u0 - u1,
+            u2*q - u1, u2*q - u0, u1*q - u2, u1*q - u0, u0*q - u2, u0*q - u1)
       To:   Rational Field
       Defn: u0 |--> 3
             u1 |--> 5
             u2 |--> 7
             q |--> 11
-    sage: mQ1 = matrix({k:fQ(v) for k, v in m1.dict().items()}); mQ1
+    sage: mQ1 = matrix({k: fQ(v) for k, v in m1.dict().items()}); mQ1                   # optional - sage.modules sage.rings.finite_rings
     [5 0 0]
     [0 3 0]
     [0 0 3]
-    sage: mQ1.base_ring()
+    sage: mQ1.base_ring()                                                               # optional - sage.modules sage.rings.finite_rings
     Rational Field
-    sage: mQ2 = matrix({k:fQ(v) for k, v in m2.dict().items()}); mQ2
+    sage: mQ2 = matrix({k: fQ(v) for k, v in m2.dict().items()}); mQ2                   # optional - sage.modules sage.rings.finite_rings
     [-15 -14   0]
     [ 26  25   0]
     [  0   0  -1]
-    sage: mQ3 = matrix({k:fQ(v) for k, v in m3.dict().items()}); mQ3
+    sage: mQ3 = matrix({k: fQ(v) for k, v in m3.dict().items()}); mQ3                   # optional - sage.modules sage.rings.finite_rings
     [    -1      0      0]
     [     0 -15/26  11/26]
     [     0 301/26 275/26]
 
     sage: S.<x, y, z, t> = QQ[]
-    sage: T = S.quo(x+y+z)
-    sage: F = T.fraction_field()
-    sage: fF = L.hom((x, y, z, t), codomain=F); fF
+    sage: T = S.quo(x + y + z)
+    sage: F = T.fraction_field()                                                        # optional - sage.libs.singular
+    sage: fF = L.hom((x, y, z, t), codomain=F); fF                                      # optional - sage.libs.singular
```

Comment:
How did you find that this code depends on `singular`? Running the code and examining the error messages on a system (p) only with the distribution that installs this file?

I wonder how we keep these tags correct in the future when new doctests are added by a PR. Will we need a github workflow that tests the branch on the system (p)?  

Could we somehow automate finding the necessary `optional` tags?

Reply:
I determined the necessary tags by running the tests of **sagemath-categories** and **sagemath-polyhedra**, inspecting the source of the errors in the displayed backtraces, and then generalizing based on what I learned. Some types of errors could be fixed by removing gratuitous dependencies (for example from module-level imports when method-level imports would suffice).

Reply:
The testsuites of **sagemath-categories** and other distributions are already being run as part of the GH Actions workflows. Currently, this does not create hard errors, but #35095 changes that.

In #35401 I have a design for a mechanism that makes maintaining the `optional` less error-prone. Not automatic, but it could be a great improvement.